docs: Plugin development guide (#2973)
vuejs/vue-cli: docs/dev-guide/plugin-dev.md

- prompting user to select certain options - for example, you can ask user if they want to create the example component mentioned above.

:::tip
Don't overuse vue-cli plugins! If you want just to include a certain dependency, e.g. [Lodash](https://lodash.com/) - it's easier to do it manually with npm than create a specific plugin only to do so.
:::

CLI Plugin should always contain a [Service Plugin](#service-plugin) as its main export, and can optionally contain a [Generator](#generator), a [Prompt File](#prompts) and a [Vue UI integration](#ui-integrtion).

As an npm package, CLI plugin must have a `package.json` file. It's also recommended to have a plugin description in `README.md` to help others find your plugin on npm.

So, typical CLI plugin folder structure looks like the following:

```
.
├── README.md
├── generator.js  # generator (optional)
├── index.js      # service plugin
├── package.json
├── prompts.js    # prompts file (optional)
└── ui.js         # Vue UI integration (optional)
```

## Generator

A Generator part of the CLI plugin is usually needed when you want to extend your package with new dependencies, create new files in your project or edit existing ones.

> Inside the CLI plugin the generator should be placed inside a `generator.js` or `generator/index.js` file. It will be invoked in two possible scenarios:

Reviewer: ```suggestion
Inside the CLI plugin the generator should be placed in a `generator.js` or `generator/index.js` file. It will be invoked in two possible scenarios:
```

- During a project's initial creation, if the CLI plugin is installed as part of the project creation preset.

> - When the plugin is installed after project's creation and invoked individually via `vue invoke`.

Reviewer: ```suggestion
- When the plugin is installed after project's creation and invoked individually via `vue add` or `vue invoke`.
```

For a 3rd party plugin, the options will be resolved from the prompts or command line arguments when the user executes `vue invoke` (see [Prompts for 3rd party plugins](#prompts-for-3rd-party-plugins)).

3. The entire preset (`presets.foo`) will be passed as the third argument.

### Creating new templates

When you call `api.render('./template')`, the generator will render files in `./template` (resolved relative to the generator file) with [EJS](https://github.com/mde/ejs).

Let's imagine we're creating [vue-cli-auto-routing](https://github.com/ktsn/vue-cli-plugin-auto-routing) plugin and we want to make the following changes to the project on plugin invoke:

- create a `layouts` folder with a default layout file;
- create a `pages` folder with `about` and `home` pages;
- add a `router.js` to the `src` folder root

To render this structure, you need to create it first inside the `generator/template` folder:

![Generator structure](/generator-template.png)

After template is created, you should add `api.render` call to the `generator/index.js` file:

```js

> api.render('./template')

Reviewer: Let's wrap this in the `module.exports = (api) => {}` function since there is no example of this before :)

- adding new vue-cli-service command - for example, `@vue/cli-plugin-unit-jest` adds a new command `test:unit` that allows developer to run unit tests;
- extending `package.json` - a useful option when your plugin adds some dependencies to the project and you need to add them to package dependencies section;
- creating new files in the project and/or modifying old ones. Sometimes it's a good idea to create an example component or modify a main file to add some imports;
- prompting user to select certain options - for example, you can ask user if they want to create the example component mentioned above.

:::tip
Don't overuse vue-cli plugins! If you want just to include a certain dependency, e.g. [Lodash](https://lodash.com/) - it's easier to do it manually with npm than create a specific plugin only to do so.
:::

CLI Plugin should always contain a [Service Plugin](#service-plugin) as its main export, and can optionally contain a [Generator](#generator), a [Prompt File](#prompts) and a [Vue UI integration](TODO).

> As a npm package, CLI plugin must have a `package.json` file. It's also recommended to have a plugin description in `README.md` to help others find your plugin on npm

Reviewer 2: ```suggestion
As an npm package, CLI plugin must have a `package.json` file. It's also recommended to have a plugin description in `README.md` to help others find your plugin on npm.
```

# Plugin Development Guide

## Getting started

> A CLI plugin is an npm package that can add additional features to a `@vue/cli` project. It should always contain a [Service Plugin](#service-plugin) as its main export, and can optionally contain a Generator, a Prompt File and a Vue UI integration.

Reviewer 3: > a `@vue/cli` project

What would you think about "a project using Vue CLI" instead? My thinking is to reference "Vue CLI" as an ecosystem of projects rather than `@vue/cli` as a package, since `@vue/cli` may not have actually been used to create the project in cases where someone upgraded or used a boilerplate - plus, `@vue/cli` will not be listed as a dependency in the project itself, which may cause confusion.

Specifying "a project using Vue CLI" instead of "a Vue CLI project" also differentiates between plugins that _tie into_ Vue CLI and applications/libraries that _make use of_ the Vue CLI ecosystem.

> It should always contain a [Service Plugin](#service-plugin) as its main export, and can optionally contain a Generator, a Prompt File and a Vue UI integration.

At a high level, what would you think about starting with user goals rather than implementation details? We're getting into the "how" of a plugin when we still haven't really answered the "why". As a user, I might be wondering, "How do I know I want to write a plugin? What are plugins for?" This paragraph contains _hints_ that I can receive options from the user, add/update project files, and add to the Vue CLI UI, but that's only the tip of the iceberg. Plus, it doesn't give me an idea of what a plugin _isn't_ appropriate for. For example, Vuetify has a plugin (because they want to make some changes to the Webpack config). Does this mean all component libraries need a plugin? Would I create a plugin to include lodash in a project?

> Generator, a Prompt File, and a Vue UI integration

Lists like this are a great place for links, for users who want to jump directly to one of these topics. If you're not sure what you want to yet or you don't want to break your flow while writing, I often just make the link `TODO` so I don't forget to come back to it (e.g. `[Generator](TODO)`).

<%# END_REPLACE %>
```

### Extending package

If you need to add an additional dependency to the project, create a new npm script or modify `package.json` in any other way, you can use API `extendPackage` method.

```js
// generator/index.js

module.exports = api => {
  api.extendPackage({
    dependencies: {
      'vue-router-layout': '^0.1.2'
    },

>   });

Reviewer: Missing end `}`

```

> In the example above we added one dependency: `vue-router-layout`. During the plugin invocation this npm module will be installed and this dependency will be added to `package.json` file.

Reviewer: ```suggestion
In the example above we added one dependency: `vue-router-layout`. During the plugin invocation this npm module will be installed and this dependency will be added to the user `package.json` file.
```

> With the same API method we can add new npm tasks to the project. To do so, we need to specify task name and a command that should be run in the `scripts` section:

Reviewer: ```suggestion
With the same API method we can add new npm tasks to the project. To do so, we need to specify task name and a command that should be run in the `scripts` section of the user `package.json`:
```

```js
// generator/index.js

module.exports = api => {
  api.extendPackage({
    scripts: {
      greet: 'vue-cli-service greet'
    }

>   });

Reviewer: Missing end `}`

```

> In the example above we're adding a new `greet` task to run a custom vue-cli service command created in [Service section](#add-a-new-cli-service-command)

Reviewer: ```suggestion
In the example above we're adding a new `greet` task to run a custom vue-cli service command created in [Service section](#add-a-new-cli-service-command).
```

### Changing main file

With generator methods you can make changes to the project files. The most usual case is some modifications to `main.js` or `main.ts` file: new imports, new `Vue.use()` calls etc.

Let's consider the case where we have created a `router.js` file via [templating](#creating-new-templates) and now we want to import this router to the main file. We will use two Generator API methods: `entryFile` will return the main file of the project (`main.js` or `main.ts`) and `injectImports` serves for adding new imports to this file:

```js
// generator/index.js

api.injectImports(api.entryFile, `import router from './router'`)
```

> Now, when we have a router imported, we can inject this router to the Vue instance in the main file. We will use `onCreateComplete` hook which is to be called when the files have been written to disk

Reviewer: ```suggestion
Now, when we have a router imported, we can inject this router to the Vue instance in the main file. We will use `onCreateComplete` hook which is to be called when the files have been written to disk.
```

}
```

In the example above we retrieve the `serve` command from the list of existing commands; then we modify its `fn` part (`fn` is the third parameter passed when you create a new command; it specifies the function to run when running the command). With the modification done the console message will be printed after `serve` command has run successfully.

### Specifying Mode for Commands

If a plugin-registered command needs to run in a specific default mode, the plugin needs to expose it via `module.exports.defaultModes` in the form of `{ [commandName]: mode }`:

``` js
module.exports = api => {
  api.registerCommand('build', () => {
    // ...
  })
}

module.exports.defaultModes = {
  build: 'production'
}
```

This is because the command's expected mode needs to be known before loading environment variables, which in turn needs to happen before loading user options / applying the plugins.

## Prompts

Prompts are required to handle user choices when creating a new project or adding a new plugin to the existing one. All prompts logic is stored inside the `prompts.js` file. The prompts are presented using [inquirer](https://github.com/SBoudrias/Inquirer.js) under the hood.

> ### Prompts for Built-in Plugins

Reviewer: Yes! 😸

> ### Prompts for Built-in Plugins

Author: Done!

![Refresh plugin](/ui-plugin-refresh.png)

## UI Integration

Vue CLI has a great [UI tool](https://github.com/vuejs/ui) which allows user to scaffold and manage a project with a nice graphical interface. The Vue CLI plugin can be integrated to this interface. UI provides an additional functionality to CLI plugins:

- you can run npm tasks, including plugin-specific ones, directly from the UI;
- you can display custom configurations for your plugin. For example, [vue-cli-plugin-apollo](https://github.com/Akryum/vue-cli-plugin-apollo) provides the following configuration screen for Apollo server:

![UI Configuration Screen](/ui-configuration.png)
- when creating the project, you can display [prompts](#prompts) visually
- you can add localizations for your plugin if you want to support multiple languages
- you can make your plugin discoverable in the Vue UI search

All the logic connected to Vue UI should be placed to `ui.js` file in the root folder or in the `ui/index.js`. The file should export a function which gets the api object as argument:

```js
module.exports = api => {
  // Use the API here...
}
```

### Augment the task in the UI

> Vue CLI plugin allows you not only add new npm tasks to the project [via Generator](#extending-package) but also create a view fro them in Vue UI. It's useful when you want to run the the task right from the UI and see it's output there.

Reviewer: ```suggestion
Vue CLI plugin allows you not only add new npm tasks to the project [via Generator](#extending-package) but also create a view for them in Vue UI. It's useful when you want to run the the task right from the UI and see its output there.
```

By default, a configuration UI might read and write to the following file types: `json`, `yaml`, `js`, `package`. Let's name our new file `myConfig.js` and place in it the root of `template` folder:

```
.
└── generator
    ├── index.js
    └── template
        ├── myConfig.js
        └── src
            ├── layouts
            ├── pages
            └── router.js
```

Now you need to add some actual config to this file:

```js
//myConfig.js

module.exports = {
  color: 'black',
};
```

After your plugin is invoked, the `myConfig.js` file will be rendered in the project root directory. Now let's add a new configuration screen with the `api.describeConfig` method in the `ui.js` file:

First you need to pass some information:

```js
//ui.js

api.describeConfig({
  // Unique ID for the config
  id: 'myConfig',
  // Displayed name
  name: 'Greeting configuration',
  // Shown below the name
  description: 'This config defines the color of the greeting printed',
  // "More info" link
  link: 'https://cli.vuejs.org/dev-guide/ui-api.html#ui-api'
})
```

#### Config logo

> You can also select an icon for your config. It can be either a [Material icon](https://material.io/tools/icons/?style=baseline) code or a custom image (see Public static files).

Reviewer: Maybe add a link on `Public static files`?

```js
api.describeConfig({
  /* ... */
  onWrite: ({ prompts, api }) => {
    // ...
  }
})
```

`onWrite` hook can take a lot of [arguments](ui-api.html#save-config-changes) but we will need only two of them: `prompts` and `api`. First one is current prompts runtime objects - we will get a prompt id from it and retrieve an answer with this id. To retrieve the answer we'll use `async getAnswer()` method from the `api`:

```js
async onWrite({ api, prompts }) {
  const result = {};
  for (const prompt of prompts) {
    result[`${prompt.id}`] = await api.getAnswer(prompt.id);
  }
  api.setData('myConfig', result);
}
```

Now if you try to change the value in the color input field from `black` to `red` on the config screen and press `Save the changes`, you will observe that `myConfig.js` file in your project has been changed as well:

```js
//myConfig.js

module.exports = {
  color: 'red',
};
```

### Display prompts

If you want, you can display [prompts](#prompts) in the Vue UI as well. If you will install your plugin through the UI, prompts will be shown on the plugin invocation step.

> To display a prompt in the UI you need to extend the [inquirer object](#prompts-for-3rd-party-plugins) with additional properties. They are optional and only used by the UI:

Reviewer: Actually you don't have to. Prompts will still be displayed anyway. However you can "augment" them for the UI with the UI-specific properties.

```js
//prompts.js

module.exports = [
  {

>     //basic propmpt properties

Reviewer: ```suggestion
    // basic prompt properties
```

    name: `addExampleRoutes`,
    type: 'confirm',
    message: 'Add example routes?',
    default: false,

>     //UI-related propmpt properties

Reviewer: ```suggestion
    //UI-related prompt properties
```

    group: 'Strongly recommended',
    description: 'Adds example pages, layouts and correct router config',
    link:
      'https://github.com/ktsn/vue-cli-plugin-auto-routing/#vue-cli-plugin-auto-routing'
  }
];
```

As a result, you will have this screen on plugin invocation:

![UI Prompts](/ui-prompts.png)

### Logo

You can put a `logo.png` file in the root directory of the folder that will be published on npm. It will be displayed in several places:
 - When searching for a plugin to install

>  - In the installed plugin list

Reviewer: - In the configuration
```suggestion
 - In the installed plugin list
 - In the configurations list (by default)
 - In the tasks list for augmented tasks (by default)
```

![Plugins](/plugins.png)

The logo should be a square non-transparent image (ideally 84x84).

### Discoverability

> For a CLI plugin to be usable by other developers, it must be published on npm following the name convention `vue-cli-plugin-<name>`. Following the name convention allows your plugin to be:

Reviewer: Or `@scope/vue-cli-plugin-<name>`
